Mathlib/FieldTheory/SeparableDegree.lean

```diff
@@ -1,147 +1,179 @@
-
-/-- In exponential characteristic one, the separable degree equals the degree. -/
-theorem HasSeparableContraction.eq_degree {f : F[X]} (hf : HasSeparableContraction 1 f) :
-    hf.degree = f.natDegree := by
-  let ⟨a, ha⟩ := hf.dvd_degree'
-  rw [← ha, one_pow a, mul_one]
-#align polynomial.has_separable_contraction.eq_degree Polynomial.HasSeparableContraction.eq_degree
-
-end CommSemiring
-
-section Field
-
-variable {F : Type*} [Field F]
-
-variable (q : ℕ) {f : F[X]} (hf : HasSeparableContraction q f)
-
-/-- Every irreducible polynomial can be contracted to a separable polynomial.
-https://stacks.math.columbia.edu/tag/09H0 -/
-theorem Irreducible.hasSeparableContraction (q : ℕ) [hF : ExpChar F q] (f : F[X])
-    (irred : Irreducible f) : HasSeparableContraction q f := by
-  cases hF
-  · exact ⟨f, irred.separable, ⟨0, by rw [pow_zero, expand_one]⟩⟩
-  · rcases exists_separable_of_irreducible q irred ‹q.Prime›.ne_zero with ⟨n, g, hgs, hge⟩
-    exact ⟨g, hgs, n, hge⟩
-#align irreducible.has_separable_contraction Polynomial.Irreducible.hasSeparableContraction
-
-/-- If two expansions (along the positive characteristic) of two separable polynomials `g` and `g'`
-agree, then they have the same degree. -/
-theorem contraction_degree_eq_or_insep [hq : NeZero q] [CharP F q] (g g' : F[X]) (m m' : ℕ)
-    (h_expand : expand F (q ^ m) g = expand F (q ^ m') g') (hg : g.Separable) (hg' : g'.Separable) :
-    g.natDegree = g'.natDegree := by
-  wlog hm : m ≤ m'
-  · exact (this q hf g' g m' m h_expand.symm hg' hg (le_of_not_le hm)).symm
-  obtain ⟨s, rfl⟩ := exists_add_of_le hm
-  rw [pow_add, expand_mul, expand_inj (pow_pos (NeZero.pos q) m)] at h_expand
-  subst h_expand
-  rcases isUnit_or_eq_zero_of_separable_expand q s (NeZero.pos q) hg with (h | rfl)
-  · rw [natDegree_expand, natDegree_eq_zero_of_isUnit h, zero_mul]
-  · rw [natDegree_expand, pow_zero, mul_one]
-#align polynomial.contraction_degree_eq_or_insep Polynomial.contraction_degree_eq_or_insep
-
-/-- The separable degree equals the degree of any separable contraction, i.e., it is unique. -/
-theorem IsSeparableContraction.degree_eq [hF : ExpChar F q] (g : F[X])
-    (hg : IsSeparableContraction q f g) : g.natDegree = hf.degree := by
-  cases hF
-  · rcases hg with ⟨_, m, hm⟩
-    rw [one_pow, expand_one] at hm
-    rw [hf.eq_degree, hm]
-  · rcases hg with ⟨hg, m, hm⟩
-    let g' := Classical.choose hf
-    obtain ⟨hg', m', hm'⟩ := Classical.choose_spec hf
-    haveI : Fact q.Prime := ⟨by assumption⟩
-    refine contraction_degree_eq_or_insep q g g' m m' ?_ hg hg'
-    rw [hm, hm']
-#align polynomial.is_separable_contraction.degree_eq Polynomial.IsSeparableContraction.degree_eq
+universe u v v' w
 
-end Field
+variable (F : Type u) (E : Type v) [Field F] [Field E]
+
+variable [Algebra F E]
 
-end Polynomial
+namespace Field
+
+/-- `Emb F E` is the type of `F`-algebra homomorphisms from `E` to the algebraic closure of `E`.
+-/
+def Emb := E →ₐ[F] (AlgebraicClosure E)
```

Comment:
I'm not sure how much you really need to talk about `Emb F E` as opposed to `E →ₐ[F] K` for a general (or algebraically closed) `K`. It would be of some general interest to prove results about such AlgHoms.

I think one thing we should show is that the cardinality of `E →ₐ[F] K` is independent of `K` as long as the `F`-minimal polynomial of every element of `E` splits in `K`. Quite a few results in mathlib are stated with such conditions, e.g. [Polynomial.lift_of_splits](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Adjoin/Field.html#Polynomial.lift_of_splits). We could show this by showing that `E →ₐ[F] K` is canonically in bijection with `E →ₐ[F] normalClosure F E K`, and that the `normalClosure` in two different `K`s are isomorphic as long as `K` satisfies the splitting condition. 

Regarding
> Mathematically, it should be the algebraic closure of F, but in order to make the type compatible with Module.rank F E, we use the algebraic closure of E.

I don't quite understand the difficulty of working with `AlgebraicClosure F`. I think it would make the inductive step to show multiplicativity of separable degree slightly easier.

Reply:
Addendum: I think we should define `IsNormalClosure F E K` to mean that (1) `K` is normal over `F`, and (2) `K` is generated by roots of minimal polynomials of elements of `E` over `F`. This is equivalent to (2) + (3) every minimal polynomial of an element of `E` over `F` splits in `K`. A stronger condition (4), that every element of `K` shares minimal polynomial with some element of `E`, also holds, and make it easy to see (1) and (3) are equivalent. We then show that `IsNormalClosure F E (normalClosure F E K)` if (3) holds, and that any two `K` that satisfy `IsNormalClosure F E K` are `F`-AlgEquiv (similar to [Polynomial.IsSplittingField.algEquiv](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/SplittingField/Construction.html#Polynomial.IsSplittingField.algEquiv)). A way to show the latter is to embed one `K` into the algebraic closure of the other `K`, and show the image must lie within the other `K`. Such an AlgHom also exists in the reverse direction, which allows you to conclude both AlgHoms are bijective using a proof similar to [Algebra.IsAlgebraic.bijective_of_isScalarTower](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Algebraic.html#Algebra.IsAlgebraic.bijective_of_isScalarTower).

Reply:
> I'm not sure how much you really need to talk about `Emb F E`

Because I want to make the definition of `sepDegree` depends only on `F` and `E`. Originally it was
```lean
def sepDegree := Cardinal.mk (E →ₐ[F] (AlgebraicClosure E))
```
but then I need to copy `E →ₐ[F] (AlgebraicClosure E)` over and over again. I saw the notation `Emb_F(E)` in several textbooks, so I propose the notation `Emb F E`.

> ... as opposed to `E →ₐ[F] K` for a general (or algebraically closed) `K`. It would be of some general interest to prove results about such AlgHoms.
>
> I think one thing we should show is that the cardinality of `E →ₐ[F] K` is independent of `K` as long as the `F`-minimal polynomial of every element of `E` splits in `K`. Quite a few results in mathlib are stated with such conditions, e.g. [Polynomial.lift_of_splits](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Adjoin/Field.html#Polynomial.lift_of_splits). We could show this by showing that `E →ₐ[F] K` is canonically in bijection with `E →ₐ[F] normalClosure F E K`, and that the `normalClosure` in two different `K`s are isomorphic as long as `K` satisfies the splitting condition. 

You're right. Currently I only have the (messy) proof of the following result:
```lean
def emb_equiv_of_isAlgClosed (halg : Algebra.IsAlgebraic F E)
    (K : Type w) [Field K] [Algebra F K] [IsAlgClosed K] :
    Emb F E ≃ (E →ₐ[F] K) := by ...
```
The `[IsAlgClosed K]` could be replaced by the condition that the `F`-minimal polynomial of every element of `E` splits in `K`.

> Regarding
> 
> > Mathematically, it should be the algebraic closure of F, but in order to make the type compatible with Module.rank F E, we use the algebraic closure of E.
> 
> I don't quite understand the difficulty of working with `AlgebraicClosure F`. I think it would make the inductive step to show multiplicativity of separable degree slightly easier.

Because the type of `Cardinal.mk (E →ₐ[F] (AlgebraicClosure F))` is `Cardinal.{max u v}`, while the type of `Cardinal.mk (E →ₐ[F] (AlgebraicClosure E))` is `Cardinal.{v}`, which is the same as `Module.rank F E`. This allows us to write `sepDegree F E = Module.rank F E` without mentioning `Cardinal.lift`.

> Addendum: I think we should define a predicate `IsNormalClosure F E K` to mean that (1) `K` is normal over `F`, and (2) `K` is generated by roots of minimal polynomials of elements of `E` over `F`. This is equivalent to (2) + (3) every minimal polynomial of an element of `E` over `F` splits in `K`. ~A stronger condition (4), that every element of `K` shares minimal polynomial with some element of `E`, also holds, and make it easy to see (1) and (3) are equivalent. (this is wrong)~ We then show that `IsNormalClosure F E (normalClosure F E K)` if (3) holds, and that any two `K` that satisfy `IsNormalClosure F E K` are `F`-AlgEquiv (similar to [Polynomial.IsSplittingField.algEquiv](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/SplittingField/Construction.html#Polynomial.IsSplittingField.algEquiv)). A way to show the latter is to embed one `K` into the algebraic closure of the other `K`, and show the image must lie within the other `K`. Such an AlgHom also exists in the reverse direction, which allows you to conclude both AlgHoms are bijective using a proof similar to [Algebra.IsAlgebraic.bijective_of_isScalarTower](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Algebraic.html#Algebra.IsAlgebraic.bijective_of_isScalarTower).

Maybe `IsNormalClosed F E K` (or more precisely, `IsSplitIn F E K`) is enough, i.e. we allow `K` to be larger than the normal closure.

Unrelated: I found the capital letter inconsistency: `AlgebraicClosure` is PascalCase while `normalClosure` is camelCase. Does that have a reason?

Reply:
> Because the type of Cardinal.mk (E →ₐ[F] (AlgebraicClosure F)) is Cardinal.{max u v}, while the type of Cardinal.mk (E →ₐ[F] (AlgebraicClosure E)) is Cardinal.{v}

I see!

> Maybe IsNormalClosed F E K (or more precisely, IsSplitIn F E K) is enough, i.e. we allow K to be larger than the normal closure.

IsSplitIn is a single condition so I wouldn't introduce a def for it. IsNormalClosure is a combination of two conditions and hace an equivalent characterization so I'd introduce a def. Of course, IsSplitIn isn't sufficient for uniqueness (up to AlgEquiv), but we need an AlgEquiv to show the embeddings into two normalClosures are of equal cardinality. I'll submit a PR adding IsNormalClosure during the weekend unless you beat me :)

Reply:
> Unrelated: I found the capital letter inconsistency: AlgebraicClosure is PascalCase while normalClosure is camelCase. Does that have a reason?

normalClosure is an IntermediateField while AlgebraicClosure is a type; only types or Props get PascalCase.

Reply:
> Of course, IsSplitIn isn't sufficient for uniqueness (up to AlgEquiv), but we need an AlgEquiv to show the embeddings into two normalClosures are of equal cardinality.

Let me think a bit. I suspect that we don't need the uniqueness to prove the equal cardinality (see the newly added `emb_equiv_of_isAlgClosed` at L288), so I think `IsSplitIn` is enough.

> I'll submit a PR adding IsNormalClosure during the weekend unless you beat me :)

Of course `IsNormalClosure` is an important concept. I'll not touch it since you self-assigned the task :)

Reply:
> Let me think a bit. I suspect that we don't need the uniqueness to prove the equal cardinality 

Yeah that's true, but *my* proof consists of two steps: first show that restriction to the `normalClosure` doesn't change the cardinality (which is obvious, since `normalClosure` is defined to be the intermediate field generated by the images of embeddings), and second show that all `normalClosure`s are the same up to AlgEquiv assuming my condition (3) (i.e. your IsSplitIn); these `normalClosure`s satisfy (2) + (3) and I'll show it also satisfies (1) (normality) using Normal.ofIsSplittingField.